### ISSUE TITLE
Low mIoU on Cityscapes

### ISSUE THREAD

**Issue by fabvio** [jfzhang95/pytorch-deeplab-xception#35]
Hi, thanks for sharing your code. I trained the model on the cityscapes dataset (btw, you missed a `self` [here](https://github.com/jfzhang95/pytorch-deeplab-xception/blob/0a5814f77fcbf3205eb77aa52c3e8793096e9f5d/dataloaders/datasets/cityscapes.py#L72)) without code edits and I can get only 68% mIoU. Do you have any pretrained models, or can you describe your training strategy? I trained on a single Tesla v100 GPU, with lr = 0.007 and batch size 8.
Thanks in advance

**Comment by jfzhang95** [comment]
What is the backbone you used in your model?
Actually I did not train the model on cityscapes dataset.

**Comment by fabvio** [comment]
The backbone is Resnet, as I have seen in another issue that Xception may have some problems. Am I right?

**Comment by jfzhang95** [comment]
Yes, previous code miss bn after depth wise cnn. I have fixed it. But I do not have pretrained aligned xception model right now. So I dont recommend you use this backbone.

**Comment by fabvio** [comment]
I think that the low performances could be related to the issue #20, because in [your code](https://github.com/jfzhang95/pytorch-deeplab-xception/blob/e08295cef0793c16aafcffe4d04ca446c2387b4e/dataloaders/custom_transforms.py#L111) you fill to 0, and not to `ignore_index`, causing the network to learn street features on the black part of the image.

**Comment by jfzhang95** [comment]
Hi, it is an error in `previous` branch, but in current branch, I have already fixed it.

**Comment by fabvio** [comment]
Are you sure abuot that? In the cityscapes dataset [here](https://github.com/jfzhang95/pytorch-deeplab-xception/blob/b7cbf5a5ae8b3197ce9117f8fc64e0324bce4b3b/dataloaders/datasets/cityscapes.py#L84) you are referring to the method I linked before, with 0 filling, so you are actually telling your network that all the padding you have belongs to a class. This is on `master`, tell me if I got something wrong.

**Comment by jfzhang95** [comment]
Hi, I think you are right. 
When training model on cityscapes dataset, [mask](https://github.com/jfzhang95/pytorch-deeplab-xception/blob/b7cbf5a5ae8b3197ce9117f8fc64e0324bce4b3b/dataloaders/custom_transforms.py#L112) should be padded to 255 instead of 0. Otherwise it will affect final performance.

**Comment by fabvio** [comment]
I will test the model trained with the updated code and report the performances to you. Right now, I am training with really resized images (512x256), batch size 16 and lr=0.007.

**Comment by fabvio** [comment]
Update on the results: very low mIoU (around 0.6). I think there is some additional problem with the code, because I trained also on another dataset (bdd100k) and the results are equally bad. Or maybe my training parameters are wrong, but they are similar to yours...

**Comment by shiretzet** [comment]
Same here... low mIoU (around 0.6), base_size = 1024, crop=600

**Comment by jfzhang95** [comment]
I tried to train ResNet based model on cityscapes dataset. My training setting is batch_size 16 (8 TITAN XP GPUs), base_size and crop_size 796, lr 0.01, epoch 200. Currently on epochs 50, and best mIoU is 0.65. So I think we need to tune training parameters in order to reach higher mIoU.

**Comment by jfzhang95** [comment]
Maybe you could explore [pytorch-encoding](
https://github.com/zhanghang1989/PyTorch-Encoding) or [drn](
https://github.com/fyu/drn) repos to try their training setting and data augmentation methods on Cityscapes dataset. I believe their code could help you.

**Comment by sunguanxiong** [comment]
Same Low IOU on cityscapes, 200 epoch, 8 GPUs, BS = 16, lr 0.01, base_size 1024, crop_size 769. The Final result is 67.98%

**Comment by jfzhang95** [comment]
Hi, all

I just noticed an [issue](https://github.com/sacmehta/ESPNet/issues/9#issuecomment-392057816) in [ESPNet](https://github.com/sacmehta/ESPNet/). It seems like Cityscapes dataset script uses weighted mIoU. So maybe you can evaluate mIoU using the official script instead of script implemented in this repo.

**Comment by wyfeng1020** [comment]
But the use of whether general or weighted mIou should not cause such a large performance gap.

**Comment by herleeyandi** [comment]
@jfzhang95 @fabvio  How to do a single inference?, I have train in cityscapes I got only 62% mIoU. After I am doing testing I got this result. Why the result seems very bad? I have doing the normalization first before feeding to the network like substracted mean and unit variance.
![image](https://user-images.githubusercontent.com/2309538/51274375-fe995400-1a09-11e9-986c-c96dce94afdc.png)

**Comment by yazhe2017** [comment]
Hi, thanks for the great repo. I try to train the cityscapes dataset with the suggested parameters. However, the performance of network is weird. The epoch loss decrease all the way. However, it is weird to note that the miou first rise then descend and the highest miou is 57.5%. Could you please try to give a suggestion on this? Thanks a lot.
My parameters setting is 4 v100 GPUs, bs=16, base_size=513, crop_size=513, backbone=resnet lr=0.01. I made additional trying with setting as 8 v100 GPUs, base_size=1024, crop_size=769. However, the performances for both experiments are similar. 
By the way, I use the pytorch provided model (Imagenet) as the pretrained backbone.

![loss](https://user-images.githubusercontent.com/34027393/52681668-9d689000-2f77-11e9-9804-10c38e857e50.png)

![miou](https://user-images.githubusercontent.com/34027393/52681673-a194ad80-2f77-11e9-9f78-3f9e70859e59.png)

**Comment by GeoffreyChen777** [comment]
This repo use nn.Dropout instead of nn.Dropout2d. Maybe this is one reason.

**Comment by EchoAmor** [comment]
> This repo use nn.Dropout instead of nn.Dropout2d. Maybe this is one reason.

Hello ,have u solve this problem yet? can you share your results? I also only got 68% mIOU,and dont know how to improve . Thank you very much!

**Comment by yazhe2017** [comment]
I found the problem for my case probably caused by the function of SynchronizedBN. When I turn of the SynBN and use the nn.BatchNorm2d, I have the highest performance 72% in val dataset without coco pretrained. However, it is weird to note that the miou descends in the middle of training. By the way, my setting is  bs=4, base_size=796, crop_size=796, backbone=resnet lr=0.01 1GPU(v100). The current miou is still far away from the state-of-the-art performance, maybe there are other problems.

![screenshot from 2019-02-27 14-50-10](https://user-images.githubusercontent.com/34027393/53471143-b5bfcb00-3a9e-11e9-8137-65afdabc29e5.png)

**Comment by sreeragh-ar** [comment]
@jfzhang95 
> Hi, all
> 
> I just noticed an [issue](https://github.com/sacmehta/ESPNet/issues/9#issuecomment-392057816) in [ESPNet](https://github.com/sacmehta/ESPNet/). It seems like Cityscapes dataset script uses weighted mIoU. So maybe you can evaluate mIoU using the official script instead of script implemented in this repo.


Trained the model on Cityscapes for 100 epochs, evaluated using official Cityscapes script. Obtained mIoU of **0.694**.

**Evaluation using the repo**
![image](https://user-images.githubusercontent.com/19902556/103234146-e60b7580-4964-11eb-9bec-c5e9070edab8.png)
The best mIoU obtained was **0.6816**.

I used the **whole image** i.e, 1024*2048 during **evaluation**. The mIoU improved slightly to **0.6943**

**Evaluation using official Cityscapes scripts**
Installed
 `pip install cityscapesScripts`.
Saved the predictions (used input size 1024*2048) as images (after converting train ids to valid class ids) in a folder.
And used `csEvalPixelLevelSemanticLabeling` command.

The mIoU obtained was **0.694**
![image](https://user-images.githubusercontent.com/19902556/103234740-68486980-4966-11eb-94d3-a973a20e1cca.png)

**Comment by ShunZuo-AI** [comment]
> What is the backbone you used in your model? Actually I did not train the model on cityscapes dataset.

the metrics Acc np.diag(self.confusion_matrix) / self.confusion_matrix.sum(axis=1) may be wrong,please check it.
